File: src/Player.cpp
#include "Player.h"
using namespace std;

Player::Player(int arg_id, string arg_name) {
	id = arg_id;
	name = arg_name;
	score = 0;
}

int  Player::get_id() const{
	return id;
}

std::string Player::get_name() const{
	return name;
}

int Player::get_score() const{
	return score;
}

Tile Player::get_top() const {
    return tiles.top();
}

int Player::get_nb_tiles() const{
	return tiles.size();
}

void Player::push_tile(Tile t){
	tiles.push(t);
}

Tile Player::pop_tile() {
    Tile t = tiles.top();
	tiles.pop();
    return t;
}
	
ostream &operator<<(ostream &os, Player const &p) { 
    os << "[id:" << p.get_id() << ", " << p.get_name() << "] " << p.get_nb_tiles() << " tiles. ";
    if (p.get_nb_tiles() != 0)
		os << "Last tile : number = " << p.get_top().get_number() << "; value = " << p.get_top().get_value();
	return os;
}

File: src/Player.java
import java.util.Stack;

class Player {
    private int id;
    private int score;
    private String name;
    private Stack<Tile> tileStack;
    private Intelligence intelligence;

    private static int playerCount = 0;

    public Player(String name, Intelligence intelligence) {
        this.id = playerCount++;
        this.score = 0;
        this.name = name;
        this.tileStack = new Stack<Tile> ();
        this.intelligence = intelligence;
    }

    public Player(Intelligence intelligence) {
        this("Player " + playerCount, intelligence);
    }

    public int getId() {
        return id;
    }

    public int getScore() {
        return score;
    }

    public String getName() {
        return name;
    }

    public boolean equals(Object o) {
        return id == ((Player)o).getId();
    }

    public Tile[] getTileStack() {
        return (Tile[])(tileStack.toArray());
    }

    public Intelligence getIntelligence() {
        return intelligence;
    }

    public Tile getTopTile() {
        if(tileStack.empty()) 
            return null;
        else
            return tileStack.peek();
    }

    public Tile popTile() {
        if(tileStack.empty())
            return null;
        else {
            Tile t = tileStack.pop();
            score -= t.getNbWorms();
            return t;
        }
    }

    public void pushTile(Tile t) {
        tileStack.push(t);
        score += t.getNbWorms();
    }
}

